### PR TITLE
[RLlib] Replay Buffer API and Ape-X

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

- Introduce Replay Buffer API to Ape-X

- Deprecate buffers that are not needed anymore, since Ape-X is the last Q-Learning Algo to receive support for new API
- Clean up configs of remaining artefacts that stem from old buffer usage
- Clean up examples and tests

- Only IMPALA/APPO and their MixIn ReplayBuffer remaining

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/rllib_tests/learning_tests/hard_learning_tests.yaml [hunk 1]
```diff
@@ -240,24 +241,24 @@ ddpg-hopperbulletenv-v0:
         min_sample_timesteps_per_reporting: 1000
         target_network_update_freq: 0
         tau: 0.001
-        buffer_size: 10000
-        prioritized_replay: True
-        prioritized_replay_alpha: 0.6
-        prioritized_replay_beta: 0.4
-        prioritized_replay_eps: 0.000001
+        replay_buffer_config:
+          capacity: 10000
```
Comment (sven1977) [inline release/rllib_tests/learning_tests/hard_learning_tests.yaml:245]: Should we add type here either way? Just to make it more clear which class is used?

File: release/rllib_tests/learning_tests/hard_learning_tests.yaml [hunk 2]
```diff
@@ -311,20 +312,21 @@ dqn-breakoutnoframeskip-v4:
         dueling: false
         num_atoms: 1
         noisy: false
-        prioritized_replay: false
+        replay_buffer_config:
+          type: MultiAgentReplayBuffer
+          learning_starts: 20000
+          capacity: 1000000
+          prioritized_replay_alpha: 0.5
```
Comment (sven1977) [inline release/rllib_tests/learning_tests/hard_learning_tests.yaml:319]: Also here, comment, why we are specifying `prioritized_replay_alpha` if we don't even use the prio-replay buffer.
Probably save to remove the `prioritized_replay_alpha` altogether.

File: rllib/BUILD [hunk 3]
```diff
@@ -1030,7 +1030,7 @@ py_test(
         "--env", "Pendulum-v1",
         "--run", "APEX_DDPG",
         "--stop", "'{\"training_iteration\": 1}'",
-        "--config", "'{\"framework\": \"tf\", \"num_workers\": 2, \"optimizer\": {\"num_replay_buffer_shards\": 1}, \"learning_starts\": 100, \"min_time_s_per_reporting\": 1, \"batch_mode\": \"complete_episodes\"}'",
+        "--config", "'{\"framework\": \"tf\", \"num_workers\": 2, \"optimizer\": {\"num_replay_buffer_shards\": 1}, \"replay_buffer_config\": {\"learning_starts\": 100}, \"min_time_s_per_reporting\": 1, \"batch_mode\": \"complete_episodes\"}'",
```
Comment (sven1977) [inline rllib/BUILD:1033]: nice!
Just a note here: These tests should be deprecated soon :)

File: rllib/utils/replay_buffers/utils.py [hunk 4]
```diff
@@ -145,190 +138,119 @@ def validate_buffer_config(config: dict):
...
+            old="config['burn_in']",
+            help="config['replay_buffer_config']['replay_burn_in']",
         )
-        config["replay_buffer_config"]["capacity"] = capacity
 
     # Deprecation of old-style replay buffer args
     # Warnings before checking of we need local buffer so that algorithms
     # Without local buffer also get warned
-    deprecated_replay_buffer_keys = [
+    keys_with_deprecated_positions = [
```
Comment (sven1977) [inline rllib/utils/replay_buffers/utils.py:177]: Nice!

File: rllib/utils/replay_buffers/replay_buffer.py [hunk 5]
```diff
@@ -96,11 +119,17 @@ def __init__(
...
+                yield self.replay()
+
+        ParallelIteratorWorker.__init__(self, gen_replay, False)
+
     def __len__(self) -> int:
         """Returns the number of items currently stored in this buffer."""
         return len(self._storage)
 
-    @ExperimentalAPI
+    @DeveloperAPI
```
Comment (sven1977) [inline rllib/utils/replay_buffers/replay_buffer.py:132]: Very cool!

File: rllib/utils/replay_buffers/replay_buffer.py [hunk 6]
```diff
@@ -12,28 +12,51 @@
...
+        if mem_size > total_gb:
+            raise ValueError(msg)
+        elif mem_size > 0.2 * total_gb:
+            logger.warning(msg)
+        else:
+            logger.info(msg)
+
+
+@DeveloperAPI
+class ReplayBuffer(ParallelIteratorWorker):
```
Comment (sven1977) [inline rllib/utils/replay_buffers/replay_buffer.py:59]: Wait, why do these have to be `ParallelIteratorWorker`s? Can we remove this inheritance once we move completely off `execution_plan`s (only APPO remaining that uses a buffer).

File: rllib/tuned_examples/sac/atari-sac.yaml [hunk 7]
```diff
@@ -31,23 +31,19 @@ atari-sac-tf-and-torch:
         n_step: 1
         rollout_fragment_length: 1
         replay_buffer_config:
-          _enable_replay_buffer_api: False
-          type: MultiAgentReplayBuffer
+          type: MultiAgentPrioritizedReplayBuffer
           capacity: 1000000
           # How many steps of the model to sample before learning starts.
-          learning_starts: 1500
+          learning_starts: 100000
```
Comment (sven1977) [inline rllib/tuned_examples/sac/atari-sac.yaml:37]: Is this a mistake? Why change this from 1.5k to 100k?

File: rllib/agents/slateq/slateq.py [hunk 8]
```diff
@@ -230,6 +230,11 @@ class SlateQTrainer(DQNTrainer):
     def get_default_config(cls) -> TrainerConfigDict:
         return SlateQConfig().to_dict()
 
+    @override(DQNTrainer)
```
Comment (sven1977) [inline rllib/agents/slateq/slateq.py:233]: Why do we even need to override SlateQ validate_config here?
Don't we call `validate_buffer_config` in DQN as well and if yes, shouldn't DQN also validate the buffer config?

File: rllib/agents/ddpg/apex.py [hunk 9]
```diff
@@ -17,41 +20,49 @@
...
-        # This is preferred b/c the learner process should have quick
-        # access to the data from the buffer shards, avoiding network
-        # traffic each time samples from the buffer(s) are drawn.
-        # Set this to False for relaxing this constraint and allowing
-        # replay shards to be created on node(s) other than the one
-        # on which the learner is located.
-        "replay_buffer_shards_colocated_with_driver": True,
-        "learning_starts": 50000,
+        "replay_buffer_config": {
+            "capcity": 2000000,
```
Comment (sven1977) [inline rllib/agents/ddpg/apex.py:24]: typo: `capacity`

File: rllib/agents/ddpg/ddpg.py [hunk 10]
```diff
@@ -88,19 +88,22 @@
...
     # === Replay buffer ===
-    # Size of the replay buffer. Note that if async_updates is set, then
-    # each worker will have a replay buffer of this size.
-    "buffer_size": DEPRECATED_VALUE,
     "replay_buffer_config": {
-        "_enable_replay_buffer_api": True,
         "type": "MultiAgentPrioritizedReplayBuffer",
         "capacity": 50000,
+        # Specify prioritized replay by supplying a buffer type that supports
+        # prioritization
```
Comment (sven1977) [inline rllib/agents/ddpg/ddpg.py:95]: ... ", for example: MultiAgentPrioritizedReplayBuffer".

File: rllib/agents/ddpg/td3.py [hunk 11]
```diff
@@ -51,15 +50,16 @@
         "target_network_update_freq": 0,
         "num_workers": 0,
         "num_gpus_per_worker": 0,
-        "worker_side_prioritization": False,
```
Comment (sven1977) [inline rllib/agents/ddpg/td3.py:54]: Cool, that we are moving this into the replay buffer config as well. Looks much cleaner now.

File: rllib/agents/ddpg/apex.py [hunk 12]
```diff
@@ -17,41 +20,49 @@
...
+            "no_local_replay_buffer": True,
+            # Specify prioritized replay by supplying a buffer type that supports
+            # prioritization
+            "prioritized_replay": DEPRECATED_VALUE,
```
Comment (smorad) [inline rllib/agents/ddpg/apex.py:28]: With your new config, is the `prioritized_replay` option still required?

File: rllib/agents/ddpg/ddpg.py (hunk 10 above)
Comment (smorad) [reply rllib/agents/ddpg/ddpg.py:95]: It seems like this is being deprecated anyways

File: rllib/agents/dqn/r2d2.py [hunk 13]
```diff
@@ -32,8 +32,10 @@
 
         # === Replay buffer ===
         "replay_buffer_config": {
-            "_enable_replay_buffer_api": True,
             "type": "MultiAgentReplayBuffer",
+            # Specify prioritized replay by supplying a buffer type that supports
+            # prioritization
+            "prioritized_replay": DEPRECATED_VALUE,
```
Comment (smorad) [inline rllib/agents/dqn/r2d2.py:38]: Is this needed here? It seems R2D2 didn't need this previously.

File: rllib/utils/replay_buffers/replay_buffer.py (hunk 6 above)
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/replay_buffer.py:59]: We still have a test that relies on replay buffers being ParallelIteratorWorkers (test_execution)!
I can absolutely remove this :)

File: rllib/tuned_examples/sac/atari-sac.yaml (hunk 7 above)
Comment (ArturNiederfahrenhorst) [reply rllib/tuned_examples/sac/atari-sac.yaml:37]: There where two `learning_starts` in the config - so I had to pick one and chose the one that was there in our last release!

File: rllib/agents/slateq/slateq.py (hunk 8 above)
Comment (ArturNiederfahrenhorst) [reply rllib/agents/slateq/slateq.py:233]: Thanks, changed!

File: rllib/agents/ddpg/apex.py (hunk 9 above)
Comment (ArturNiederfahrenhorst) [reply rllib/agents/ddpg/apex.py:24]: Thanks!

File: rllib/agents/ddpg/apex.py (hunk 12 above)
Comment (ArturNiederfahrenhorst) [reply rllib/agents/ddpg/apex.py:28]: No. I just wanted to leave it there since it's being deprecated!

File: rllib/agents/dqn/r2d2.py (hunk 13 above)
Comment (ArturNiederfahrenhorst) [reply rllib/agents/dqn/r2d2.py:38]: PER was not used in R2D2 by default, but `prioritized_replay` it was a configuration option and is not being deprecated.

File: release/rllib_tests/learning_tests/hard_learning_tests.yaml (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply release/rllib_tests/learning_tests/hard_learning_tests.yaml:319]: Ok, I thought it was there hinting "if you use PER, use it with this alpha!".
Removing it!

File: release/rllib_tests/learning_tests/hard_learning_tests.yaml (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply release/rllib_tests/learning_tests/hard_learning_tests.yaml:245]: Ok, I've added explicit types to all yamls.

**Review by sven1977 — APPROVED** [review]
Looks great! Thanks for the PR @ArturNiederfahrenhorst . Just waiting for all tests to pass again. ...

**Pull request merged**